### ISSUE TITLE
False Negative error with rust-peg macros

### ISSUE THREAD

**Issue by linkdd** [rust-lang/rust-analyzer#13013]
**rust-analyzer version**: `v0.3.1162`

**rustc version**: `rustc 1.61.0 (fe5b13d68 2022-05-18)`

**relevant settings**: ` `

---

When using the crate `rust-peg` to define a grammar, if I use their `precedence!{}` macro, **rust-analyzer** reports the following error:

```
expected &dyn Fn(usize, &mut ParseState, &mut ErrorState, &dyn Fn(usize, i32, &mut ParseState, &mut ErrorState)),
found
&|usize, &mut ParseState, {unknown}, {unknown}| -> RuleResult<Node<TypeRef>>
```

But the compiler manages to compile the code just fine.

See related (now closed) issue: https://github.com/kevinmehall/rust-peg/issues/316#issuecomment-1210809167

**Comment by flodiebold** [comment]
Can you provide a simple example to reproduce the problem?

**Comment by lowr** [comment]
Looks like the same issue as #13031. I've tested the following (incomplete) code and the error seems to go away with #13049 (`peg` does [use](https://github.com/kevinmehall/rust-peg/blob/24482eb17de16e25e19d092dc3e339a982f9b4af/peg-macros/translate.rs#L939-L940) bare dyn types).

@linkdd Would you test if #13049 fixes the issue?

<details>
<summary>code</summary>

```rust
enum Token { Comma, OperatorBinAnd, OperatorBinOr, ParenthesisBegin, ParenthesisEnd }
pub type LocationInfo = (usize, usize);

#[derive(Clone, Debug, PartialEq)]
pub struct Node<T> {
  pub location: LocationInfo,
  pub data: Box<T>,
}

peg::parser!{
    grammar list_parser() for str {
        rule type_refs() -> Vec<Node<ast::types::TypeRef>>
  = type_ref() ++ [Token::Comma]

rule type_ref() -> Node<ast::types::TypeRef>
  = precedence!{
    l:position!() data:@ r:position!() {
      Node::new((l, r), data)
    }
    --
    lhs:(@) [Token::OperatorBinOr] rhs:@ {
      ast::types::TypeRef::one_of(vec![lhs, rhs])
    }
    --
    lhs:(@) [Token::OperatorBinAnd] rhs:@ {
      ast::types::TypeRef::all_of(vec![lhs, rhs])
    }
    --
    [Token::Negation] t:(@) {
      ast::types::TypeRef::not(t)
    }
    --
    t:type_ref_term() { t }
    [Token::ParenthesisBegin] t:type_ref() [Token::ParenthesisEnd] { t.data }
  }

rule type_ref_term() -> Box<ast::types::TypeRef>
  = type_ref_term_generic_symbol()
  / type_ref_term_concrete_symbol()
  / type_ref_term_struct_def()
  / type_ref_term_tuple_def()
  / type_ref_term_function_signature()
  / type_ref_term_literal()
    }
}
```

</detail>

**Comment by linkdd** [comment]
@flodiebold Hi, sorry I was a bit sick the past few days.

@lowr Yes thank you, it works as well for me here. I'm closing this issue since it's a duplicate
